src/test/java/hudson/plugins/jira/JiraSite2Test.java

```diff
@@ -170,4 +179,137 @@ public void testDeserializeWithoutCredentials() {
+    @WithoutJenkins
+    public void ensureUrlEndsWithSlash() {
+        JiraSite jiraSite = new JiraSite(nonExistentUrl.toExternalForm());
+        jiraSite.setAlternativeUrl(exampleOrg.toExternalForm());
+        assertTrue(jiraSite.getUrl().toExternalForm().endsWith("/"));
+        assertTrue(jiraSite.getAlternativeUrl().toExternalForm().endsWith("/"));
+        URL url1 = JiraSite.toURL(nonExistentUrl.toExternalForm());
+        URL url2 = JiraSite.toURL(exampleOrg.toExternalForm());
+        assertTrue(url1.toExternalForm().endsWith("/"));
+        assertTrue(url2.toExternalForm().endsWith("/"));
+    }
+
+    @Test
+    @WithoutJenkins
+    public void urlNulls() {
+        JiraSite jiraSite = new JiraSite(nonExistentUrl.toExternalForm());
+        jiraSite.setAlternativeUrl(" ");
+        URL url1 = JiraSite.toURL("");
+        URL url2 = JiraSite.toURL(" ");
+        assertNotNull(jiraSite.getUrl());
+        assertNull(jiraSite.getAlternativeUrl());
+        assertNull(url1);
+        assertNull(url2);
+    }
+
+    @Test
+    @WithoutJenkins
+    public void ensureMainUrlIsMandatory() {
+        thrown.expect(AssertionError.class);
+        thrown.expectMessage("URL cannot be null");
+        new JiraSite("");
+    }
+
+    @Test
+    @WithoutJenkins
+    public void ensureAlternativeUrlIsNotMandatory() {
+        JiraSite jiraSite = new JiraSite(nonExistentUrl.toExternalForm());
+        jiraSite.setAlternativeUrl("");
+        assertNull(jiraSite.getAlternativeUrl());
+    }
+
+    @Test
+    @WithoutJenkins
+    public void malformedUrl() {
+        thrown.expect(AssertionError.class);
+        thrown.expectMessage("java.net.MalformedURLException: no protocol");
+        new JiraSite("malformed.url");
+    }
+
+    @Test
+    @WithoutJenkins
+    public void malformedAlternativeUrl() {
+        JiraSite jiraSite = new JiraSite(nonExistentUrl.toExternalForm());
+        thrown.expect(AssertionError.class);
+        thrown.expectMessage("java.net.MalformedURLException: no protocol");
+        jiraSite.setAlternativeUrl("malformed.url");
+    }
+
+    @Test
+    public void credentials() throws Exception {
+        JiraSite jiraSite = new JiraSite(exampleOrg.toExternalForm());
+        jiraSite.setCredentialsId("");
+        assertNull(jiraSite.getCredentialsId());
+        assertNull(jiraSite.credentials);
+        String cred = "cred-1-id";
+        String user = "user1";
+        String pwd = "pwd1";
+
+        UsernamePasswordCredentialsImpl credentials = new UsernamePasswordCredentialsImpl(
+            CredentialsScope.GLOBAL, cred, null, user, pwd);
+
+        SystemCredentialsProvider systemProvider = SystemCredentialsProvider.getInstance();
+        systemProvider.getCredentials().add(credentials);
+        systemProvider.save();
+        jiraSite.setCredentialsId(cred);
+        assertNotNull(jiraSite.getCredentialsId());
+        assertNotNull(jiraSite.credentials);
+        assertEquals(cred, jiraSite.getCredentialsId());
```

Comment:
we don't really care which ID that credentials has. so i would suggest to check username/password against raw `user` and `pwd` values.